pkg/injector/init_container.go

```diff
@@ -23,6 +24,9 @@ func getInitContainerSpec(containerName string, cfg configurator.Configurator, o
 					"NET_ADMIN",
 				},
 			},
+			RunAsNonRoot: pointer.BoolPtr(false),
```

Comment:
I don't get it after reading the description. 

Isn't `RunAsNonRoot = false` and `RunAsUser = 0` the same meaning? So setting them both at the same time solves the problem?

Reply:
![image](https://user-images.githubusercontent.com/13926417/141191465-a6777b8a-d843-4ca5-bae2-f95e9a82c870.png)


![image](https://user-images.githubusercontent.com/13926417/141191516-35f33e73-1f3e-4967-a365-bfe00b9adc2c.png)

![image](https://user-images.githubusercontent.com/13926417/141191652-08e1eff5-630b-4f0a-984c-24e493def60c.png)

Reply:
Oh man, feels hacky. But this answer solves my puzzle.

Reply:
> Setting `RunAsRoot` as `false` is redundant as it just means "do not perform such validation".

`RunAsRoot` can be set at either the pod level or the container level. Setting it here for the container is necessary in case the user specifies `runAsRoot: true` for the pod.

> Better solution would be to not touch `RunAsNonRoot` and `RunAsUser`, and instead set `Privileged` to `true`.

Privileged mode grants much higher privileges than simply running as root in an unprivileged container. Privileged doesn't seem necessary here.

This is at least my understanding. @ksubrmnn or @SanyaKochhar please correct me if I'm wrong.

Reply:
Actually just verified and you are right, using just `runAsUser = 0` is sufficient to fix the problem caused the security context originally in the description (updating now), but if that pod security context were to include `runAsNonRoot: true`, it causes the following error: 
> Error: container's runAsUser breaks non-root policy (pod: "bookstore-569cdc978b-wbd5n_bookstore(35e391ed-0d47-4066-9e5f-8b08ff974f44)", container: osm-init)

Here's why we need both to account for the different pod/container security permutations:

You're right that the nuance between the two is very slight and the naming makes it really confusing. According to k8s API docs, "If true, the Kubelet will validate the image at runtime to ensure that it does not run as UID 0 (root) and fail to start the container if it does. **If unset or false, no such validation will be performed.**"

Effectively, `runAsNonRoot: false` does not actually set the user to root, it just allows for it to be set to root. `runAsUser` additionally sets it to root user. So if the **init container** uses `runAsUser` without `runAsNonRoot: false` and the **pod** security context were to set `runAsNonRoot` to true, it would not allow `runAsUser: 0` at the **container** security context level.

Reply:
oops @nojnhuh missed your comment and didn't see that you had already addressed this, thanks!!

Reply:
Exactly @nojnhuh